### DISCUSSION TITLE
lerna bootstrap with private packages

### DISCUSSION THREAD

**Discussion by wyardley** [renovatebot/renovate#12078] [Request Help]
I read through

https://github.com/renovatebot/renovate/issues/1441
https://github.com/renovatebot/renovate/pull/1407

I have a repo with lerna used to bootstrap private packages (i.e., there's a private package in the subdirectory's `package.json`, but not 

At one point, there was an `.npmrc` checked in, and things worked fine. After removing it, renovate started throwing an artifact error on `lerna bootstrap`:
```
lerna info Installing external dependencies
lerna ERR! npm install --ignore-scripts --ignore-scripts --no-audit --package-lock-only exited 1 in '@myorg/foo'
[...]
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/@myorg%2fbar - Not found
```

Do I need a skeleton `.npmrc` checked in, or should renovate be handling this automagically? I.e., if I check one in with a literal `//registry.npmjs.org/:_authToken=${NPM_TOKEN}`, will that create it?

Other projects sharing this renovate config work fine with the encrypted token in our org level renovate config (though haven't tested other projects using lerna yet, I think). That is configured thusly:

```json
  "hostRules": [
    {
      "hostType": "npm",
      "matchHost": "registry.npmjs.org",
      "encrypted": {
        "token": "xxx"
      }
    }
  ],
```

**Comment by rarkins** [comment]
When you had a `.npmrc` checked in previously, was it including the token in plain text?

**Comment by wyardley** [comment]
> When you had a .npmrc checked in previously, was it including the token in plain text?

Yes 😞 (not by me 😉).
I tried deleting the branch etc., but let me know if there's anything cached on Renovate's side or anything else I can do to fix the issue now that it's gone.

**Reply by rarkins** [reply]
Not judging, just asking :). I was surprised if the mere presence of a `.npmrc` made such a difference, although it's always possible.

Next step - especially in lieu of a reproduction - is to code inspect whether our logic for synthesizing a `.npmrc` from `hostRules` is skipped when in lerna mode. I know it's used when in regular `npm` mode but maybe not executed when `lerna`.

Here's the function called to get the content: https://github.com/renovatebot/renovate/blob/8f81bb01656081489f94a9039c990e9e3d3fb221/lib/manager/npm/post-update/rules.ts#L10

It's called from here: https://github.com/renovatebot/renovate/blob/8f81bb01656081489f94a9039c990e9e3d3fb221/lib/manager/npm/post-update/index.ts#L414

And it does seem to be applied here: https://github.com/renovatebot/renovate/blob/8f81bb01656081489f94a9039c990e9e3d3fb221/lib/manager/npm/post-update/index.ts#L676-L680

Unless someone can do a deeper code inspection than this to find the problem, I think we'd need a reproduction repo:
- at minimum, enough to test out that the `.npmrc` is being written to the right place(s) before `lerna` is called, or
- if that's not enough, a fully working example with real private package and token.. but hopefully we don't need that

**Comment by wyardley** [comment]
@rarkins are you all able to see the logs for jobs under https://app.renovatebot.com? i.e., if I give you a job ID, will that help you see anything from the debug output?

**Reply by rarkins** [reply]
I can access them but looking through a production monorepo's logs is not fun so that's why we ask for a minimal reproduction

**Comment by wyardley** [comment]
@rarkins will https://github.com/wyardley/renovate-lerna-repro work? (no private dep in the lerna subpackage yet).

**Reply by rarkins** [reply]
I think so. Renovate should write any credentials info before calling lerna/npm regardless. Is there any indication of that in the logs?

**Reply by wyardley** [reply]
Not sure what to look for...

https://app.renovatebot.com/dashboard#github/wyardley/renovate-lerna-repro/482557576 has:
```
DEBUG: Resetting npmrc
[...]
DEBUG: Decrypted token
DEBUG: Setting hostRules from config
DEBUG: Adding token authentication for registry.npmjs.org to hostRules
[...]
```

**Reply by rarkins** [reply]
It would be branch-level logs prior to lerna being called

**Reply by wyardley** [reply]
I'm sorry, maybe I'm dense, but can you give a more specific pointer of what you want me to look for? Or could you or someone take a look at the logs for the repro repo?

**Reply by rarkins** [reply]
Renovate logs with branch name in the metadata of the log, if it's working on a branch. So pick the branch you are interested in and then filter the logs for any which include e.g. branch=renovate/foo-2.x

**Reply by wyardley** [reply]
I'm still not sure if I follow, but if I go to, e.g., https://app.renovatebot.com/dashboard#github/wyardley/renovate-lerna-repro/518683574 and then filter for branch `renovate/axios-0.x`, I don't see any `npmrc` related stuff.

Does that tell us anything? note: as mentioned before, there's no private NPM dependencies setup on this repro repo.

**Reply by rarkins** [reply]
@wyardley please post logs to a gist so that they are a lasting reference from this issue, although it's always appreciated if you can simplify/filter them first.

Is there any mention of the string "npmrc" in the entire log?

**Reply by wyardley** [reply]
Can do.
https://gist.github.com/wyardley/b2eaf027b485636d384b401071172691 for just the references to `renovate/axios-0.x`.

> Is there any mention of the string "npmrc" in the entire log?

Same as before: just
`DEBUG: Resetting npmrc`

**Comment by wyardley** [comment]
Turns out this seems to be due to the main project having the `package-lock.json` .gitignored, so I think this can be closed out

**Comment by wyardley** [comment]
I'm still getting really inconsistent results here. It seems like in some cases, the artifact will show up (even on PRs with no private deps), and in other cases, it won't. So maybe the cause was caching and some weird interaction with whether the cache comes from the project root's `package.json` or the subpackage's?

**Comment by wyardley** [comment]
I have some more debug output from the renovate run now, if it's helpful:

```
 "err": {
    "name": "ExecError",
    "cmd": "/bin/sh -c docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/myorg/myorg-services\":\"/mnt/renovate/gh/myorg/myorg-services\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -w \"/mnt/renovate/gh/myorg/myorg-services\" docker.io/renovate/sidecar bash -l -c \"install-tool node v16.17.0 && install-tool npm 6.14.17 && hash -d npm 2>/dev/null || true && install-tool lerna 5.5.1 && lerna info || echo \\\"Ignoring lerna info failure\\\" && npm install --ignore-scripts  --no-audit --package-lock-only && lerna bootstrap --no-ci --ignore-scripts -- --ignore-scripts  --no-audit --package-lock-only\"",
    "stderr": "lerna notice cli v5.5.1\nlerna notice cli v5.5.1\nlerna info Bootstrapping 6 packages\nlerna info Installing external dependencies\nlerna ERR! npm install --ignore-scripts --ignore-scripts --no-audit --package-lock-only exited 1 in '@myorg/foo-service'\nlerna ERR! npm install --ignore-scripts --ignore-scripts --no-audit --package-lock-only stderr:\nnpm ERR! code E404\nnpm ERR! 404 Not Found - GET https://**redacted**@myorg%2fpackagename - Not found\nnpm ERR! 404 \nnpm ERR! 404  '@myorg/packagename@4.6.0' is not in the npm registry.\nnpm ERR! 404 You should bug the author to publish it (or use the name yourself!)\nnpm ERR! 404 It was specified as a dependency of 'foo-service'\nnpm ERR! 404 \nnpm ERR! 404 Note that you can also install from a\nnpm ERR! 404 tarball, folder, http url, or git url.\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate-cache/others/npm/_logs/2022-09-21T14_51_20_756Z-debug.log\nlerna ERR! npm install --ignore-scripts --ignore-scripts --no-audit --package-lock-only exited 1 in '@myorg/foo-service'\n",
    "stdout": "installing v2 tool node v16.17.0\nlinking tool node v16.17.0\nnode: v16.17.0 /usr/local/bin/node\nnpm: 8.15.0  /usr/local/bin/npm\nInstalled v2 /usr/local/buildpack/tools/v2/node.sh in 6 seconds\nskip cleanup, not a docker build: e9c42137bbb6\ninstalling v2 tool npm v6.14.17\n\nadded 1 package in 12s\nnpm WARN deprecated circular-json@0.3.3: CircularJSON is in maintenance only, flatted is its successor.\nnpm WARN deprecated fs-access@2.0.0: This package is no longer relevant as Node.js 0.12 is unmaintained.\nnpm WARN deprecated request@2.88.2: request has been deprecated, see https://**redacted**@parcel/watcher@2.0.4 install /home/user/lerna/5.5.1/lib/node_modules/lerna/node_modules/@parcel/watcher\n> node-gyp-build\n\n\n> nx@14.7.8 postinstall /home/user/lerna/5.5.1/lib/node_modules/lerna/node_modules/nx\n> node ./bin/compute-project-graph\n\nnpm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.3.2 (node_modules/lerna/node_modules/chokidar/node_modules/fsevents):\nnpm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {\"os\":\"darwin\",\"arch\":\"any\"} (current: {\"os\":\"linux\",\"arch\":\"x64\"})\n\n+ lerna@5.5.1\nadded 624 packages from 369 contributors in 67.161s\n5.5.1\nInstalled v1 lerna in 70 seconds\nskip cleanup, not a docker build: e9c42137bbb6\n\n Environment info:\n\n  System:\n    OS: Linux 5.4 Ubuntu 20.04.5 LTS (Focal Fossa)\n    CPU: (1) x64 Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz\n  Binaries:\n    Node: 16.17.0 - /usr/local/bin/node\n    npm: 6.14.17 - /usr/local/bin/npm\n  Utilities:\n    Git: 2.37.3 - /usr/bin/git\n\nadded 1215 packages in 10.836s\n",

```
